new package: MOLGW

bruneval opened the pull request:
Create the spack packaging for [MOLGW](https://github.com/bruneval/molgw), a many-body perturbation theory code in Gaussian basis

bruneval committed:
molgw package.py

bruneval committed:
molgw package.py almost ready

bruneval committed:
Merge branch 'develop' of github.com:spack/spack into develop

bruneval committed:
bug fix in molgw package.py

bruneval committed:
MOLGW v3.2

bruneval committed:
black pass

bruneval committed:
duplicated line eliminated

bruneval committed:
Merge branch 'develop' of github.com:spack/spack into develop

bruneval committed:
remove FIXME in the header

bruneval committed:
Merge branch 'develop' of github.com:spack/spack into develop

bruneval committed:
add me as a maintainer

bruneval committed:
fix previous commit

bruneval committed:
sort the imports in the order spack wants

bruneval committed:
chop the too-long lines

bruneval commented:
@tgamblin sorry for mentioning you personally.
But it looks like my pull request has been forgotten for long.

It is very simple: it simply adds a new package file, it leaves all the other packages untouched, and it passes all the CI tests.

I think this is a straightforward addition.

Thanks for your time.
Fabien

alalazo commented on var/spack/repos/builtin/packages/molgw/package.py:
```diff
@@ -0,0 +1,120 @@
+            for k, v in flags.items():
+                f.write(k + "=" + v + "\n")
+
+    def build(self, spec, prefix):
+        make()
+
+    def install(self, spec, prefix):
+        make("install")
```
This shouldn't be needed (it's inherited by the base package)

bruneval replied:
Fixed!

alalazo commented on var/spack/repos/builtin/packages/molgw/package.py:
```diff
@@ -0,0 +1,120 @@
+    depends_on("libxc@5:")
+    depends_on("libcint+pypzpx+coulomb_erf")
+
+    variant("openmp", default=False, description="Build with OpenMP support")
+
+    variant("scalapack", default=False, description="Build with ScaLAPACK support")
```
(minor) Can we move variants above all the `depends_on` directives?

bruneval replied:
Fixed!

alalazo commented on var/spack/repos/builtin/packages/molgw/package.py:
```diff
@@ -0,0 +1,120 @@
+    # depends_on('hdf5', when='+hdf5')
+
+    def _get_mkl_ld_flags(self, spec):
+        mklroot = str(getenv("MKLROOT"))
```
Can't we use `spec["blas"].libs` etc. to abstract the provider?

bruneval replied:
See my next comment.

alalazo commented on var/spack/repos/builtin/packages/molgw/package.py:
```diff
@@ -0,0 +1,120 @@
+                command.extend(["-o", "gomp"])
+
+        if "+scalapack" in spec:
+            command.extend(["--cluster_library=scalapack"])
```
Similarly for `scalapack`

bruneval replied:
Dear Alalazo,

I completely agree with you in principle.

The only problem is that the spec["scalapack"].libs is not correctly set in the case of `intel-oneapi-mkl`.

Indeed, in `intel-oneapi-mkl`, the variant `+cluster` adds the linking to intelmpi only.

intel-oneapi-mkl/package.py:
```
        if "+cluster" in self.spec:
            libs.append(self._xlp64_lib("libmkl_blacs_intelmpi"))
```
This linking doesn't work with openmpi for instance.

One should fix intel-oneapi-mkl first. But I'm afraid to touch such an important package by myself.
I reported this problem in the issue #37459 one month ago...

Best,
Fabien

alalazo reviewed (changes requested):
I left a few comments. Let me know what you think.

bruneval committed:
many fixes

- variants before dependences
- eliminate useless build and install stages
- no openmp with intel-mkl was broken

bruneval committed:
after blackization

bruneval committed:
cleaning

bruneval commented:
Dear @alalazo 
I think I addressed all your comments.

Please let me know if something is still missing.
Best,
Fabien

alalazo commented on var/spack/repos/builtin/packages/molgw/package.py:
```diff
@@ -0,0 +1,121 @@
+                command.extend(["-m", "mpich2"])
+            elif "intelmpi" in spec:
+                command.extend(["-m", "intelmpi"])
+        # only works with recent python3
+        # result = run(command,capture_output=True, text=True)
+        # return result.stdout.strip()
```
```suggestion
```

alalazo commented on var/spack/repos/builtin/packages/molgw/package.py:
```diff
@@ -0,0 +1,121 @@
+        # result = run(command,capture_output=True, text=True)
+        # return result.stdout.strip()
+        result = run(command, stdout=PIPE)
+        # with open("my_machine.arch", "w") as f:
+        #    f.write("#" + " ".join(command) + "\n")
+        #    f.write("#" + result.stdout.decode(encoding="utf-8").strip() + "\n")
```
```suggestion
```

alalazo commented on var/spack/repos/builtin/packages/molgw/package.py:
```diff
@@ -0,0 +1,121 @@
+    # enforce threaded openblas when asking +openmp (and using openblas)
+    depends_on("openblas threads=openmp", when="+openmp ^openblas")
+
+    def _get_mkl_ld_flags(self, spec):
```
This method seems to belong more to some `Intel*` base class, than here. @rscohn2 Do you know if we have anything similar to this?

rscohn2 replied:
If it depends on `intel-oneapi-mkl +cluster threads=openmp` or `intel-mkl threads=openmp ^mpi` then `spec["lapack"].libs` should have everything you need.

bruneval replied:
As I wrote in issue #37459, ```intel-oneapi-mkl +cluster threads=openmp``` only works when linking against `intel-mpi`.

See the corresponding lines in `intel-oneapi-mkl/package.py`:
```
        if "+cluster" in self.spec:
            libs.append(self._xlp64_lib("libmkl_blacs_intelmpi"))
```

My coding allows the linking against openmpi.

I agree that in principle the `intel-oneapi-mkl` should be fixed first, but the issue I raised remained unanswered.

bruneval replied:
If you prefer, I can propose a pull request to `intel-oneapi-mkl`. But this package is quite central and I would need some back-up.

alalazo commented on var/spack/repos/builtin/packages/molgw/package.py:
```diff
@@ -0,0 +1,121 @@
+
+    maintainers("bruneval")
+
+    # version("rolling-release", branch="master")
```
```suggestion
```

alalazo commented on var/spack/repos/builtin/packages/molgw/package.py:
```diff
@@ -0,0 +1,121 @@
+    # version("rolling-release", branch="master")
+    version("3.2", sha256="a3f9a99db52d95ce03bc3636b5999e6d92b503ec2f4afca33d030480c3e10242")
+
+    # variants
```
```suggestion
```

alalazo commented on var/spack/repos/builtin/packages/molgw/package.py:
```diff
@@ -0,0 +1,121 @@
+    variant("openmp", default=False, description="Build with OpenMP support")
+    variant("scalapack", default=False, description="Build with ScaLAPACK support")
+
+    # dependences
```
```suggestion
```

alalazo reviewed (changes requested):
A question, mainly for @rscohn2 and a minor request to remove commented out code and redundant comments.

bruneval committed:
Update var/spack/repos/builtin/packages/molgw/package.py

Co-authored-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

bruneval committed:
Update var/spack/repos/builtin/packages/molgw/package.py

Co-authored-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

bruneval committed:
Update var/spack/repos/builtin/packages/molgw/package.py

Co-authored-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

bruneval committed:
Update var/spack/repos/builtin/packages/molgw/package.py

Co-authored-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

bruneval committed:
Update var/spack/repos/builtin/packages/molgw/package.py

Co-authored-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo reviewed (approved):
Le'ts get this recipe in and fix the intel logic in a following PR
